chore(stark-all): use ng-packagr in the build process instead of rollup

SuperITMan opened the pull request:
ISSUES CLOSED: #279, #253

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/NationalBankBelgium/stark/blob/master/CONTRIBUTING.md#-commit-message-guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[X] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Our build process is complex to maintain and not standard. 
We would like to have a look on @nrwl/nx and to do that we need to already have ng-packagr in our project.

Issues Numbers: #279, #253


## What is the new behavior?

Rollup has been removed in favor of ng-packagr. It increases a lot the speed of the build.

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

christophercr commented on packages/stark-ui/src/modules/input-mask-directives/directives/text-mask.constants.ts:
```diff
@@ -1,5 +1,6 @@
 /**
  * Util class containing some constants with predefined masks for common mask patterns to be used with the {@link StarkTextMaskDirective}.
+ * @dynamic
```
Why is this added here? Can you add a comment or a link where we should look at to understand this?

christophercr commented on packages/stark-ui/testing/package.json:
```diff
@@ -1,15 +1,12 @@
-  "esm2015": "../esm2015/testing/testing.js",
-  "fesm5": "../fesm5/testing.js",
-  "fesm2015": "../fesm2015/testing.js",
+  "ngPackage": {
```
Can we also use an `ng-package.json` file for this as with `stark-ui` so we keep consistency along all our packages?

christophercr commented on packages/stark-ui/tsconfig.json:
```diff
@@ -12,9 +12,12 @@
       "@angular/common": ["../../node_modules/@angular/common"],
       "@angular/core": ["../../node_modules/@angular/core"],
       "@angular/material": ["../../node_modules/@angular/material"],
+      "@angular/material-moment-adapter": ["./node_modules/@angular/material-moment-adapter"],
```
Is this path really needed even if this is part of the `node_modules` of `stark-ui` itself?

SuperITMan replied:
As it is the case for `stark-core`, this path is needed for the compilation of the `testing` subpackage.

christophercr commented on packages/stark-ui/tsconfig.lib.json:
```diff
@@ -0,0 +1,19 @@
+    "declarationDir": "AUTOGENERATED",
+    "inlineSourceMap": true
+  },
+  "exclude": ["node_modules", "dist", "**/*.ngfactory.ts", "**/*.shim.ts", "**/*.spec.ts"],
```
Why are `**/*.ngfactory.ts` and `**/*.shim.ts` included here? We have no such files

christophercr commented on packages/tsconfig.json:
```diff
@@ -1,35 +0,0 @@
-{
-  // this config is only used by build.sh
```
Great! Finally this odd file will be gone! 😆

christophercr commented on showcase/karma.conf.js:
```diff
@@ -12,16 +12,7 @@ const defaultKarmaConfig = require("./node_modules/@nationalbankbelgium/stark-te
-			"@angular/http": "./node_modules/@angular/common/bundles/common-http.umd.js"
-		}
-	}
+	resolve: { alias: {} }
```
Just a suggestion: can this change be on a separate PR (for example the one about simplifying the karma confs #1145 ) to keep this PR clean with just the changes needed for `ng-packagr`? That way is easier to find relevant changes in the future

christophercr commented:
I also noticed some warnings during the build process:

For stark-core:
![stark-core-build-warning](https://user-images.githubusercontent.com/5698318/61882450-e1819880-aef8-11e9-9d09-e87930d4927e.png)

And for stark-ui:
![stark-ui-build-warning](https://user-images.githubusercontent.com/5698318/61882459-e6dee300-aef8-11e9-9156-d57d29a13b0a.png)

I think we should troubleshoot these and make sure they don't cause any issue in the final packages.

SuperITMan commented on packages/stark-core/tsconfig.json:
```diff
@@ -9,31 +9,21 @@
+      "@ngx-translate/core": ["./node_modules/@ngx-translate/core"],
+      "@uirouter/angular": ["./node_modules/@uirouter/angular"],
+      "@uirouter/core": ["./node_modules/@uirouter/core"],
+      "cerialize": ["./node_modules/cerialize"],
```
All these added paths are needed when compiling `testing` package.
Without them, we get this kind of errors during the compilation: 
```shell
BUILD ERROR
dist/packages/stark-core/stark-core.ts(10,2): Error during template compile of 'StarkHttpErrorBaseImpl'
  Could not resolve cerialize relative to [object Object]..
dist/packages/stark-core/stark-core.ts(11,2): Error during template compile of 'StarkHttpErrorImpl'
  Could not resolve cerialize relative to [object Object]..
dist/packages/stark-core/stark-core.ts(9,2): Error during template compile of 'StarkHttpErrorDetailImpl'
  Could not resolve cerialize relative to [object Object]..
dist/packages/stark-core/stark-core.ts(7,12): Error during template compile of 'StarkLoggingModule'
  Could not resolve @ngrx/store relative to [object Object]..
dist/packages/stark-core/stark-core.ts(6,12): Error during template compile of 'StarkRoutingModule'
  Could not resolve @uirouter/angular relative to [object Object]..
dist/packages/stark-core/stark-core.ts(19,3): Error during template compile of 'StarkSessionModule'
  Could not resolve @ng-idle/core relative to [object Object]..
dist/packages/stark-core/stark-core.ts(20,3): Error during template compile of 'StarkSessionModule'
  Could not resolve @ng-idle/keepalive relative to [object Object]..
dist/packages/stark-core/stark-core.ts(21,3): Error during template compile of 'StarkSessionModule'
  Could not resolve @ngrx/store relative to [object Object]..
dist/packages/stark-core/stark-core.ts(22,3): Error during template compile of 'StarkSessionModule'
  Could not resolve @uirouter/angular relative to [object Object]..
dist/packages/stark-core/stark-core.ts(9,12): Error during template compile of 'StarkSettingsModule'
  Could not resolve @ngrx/store relative to [object Object]..
dist/packages/stark-core/stark-core.ts(9,76): Error during template compile of 'StarkSettingsModule'
  Could not resolve @ngrx/effects relative to [object Object]..
```

SuperITMan commented on packages/stark-core/testing/package.json:
```diff
@@ -1,15 +1,12 @@
-  "esm2015": "../esm2015/testing/testing.js",
-  "fesm5": "../fesm5/testing.js",
-  "fesm2015": "../fesm2015/testing.js",
+  "ngPackage": {
```
You're right, we can do it as this. I changed this and added the ng-package.json in every testing subpackage 😊

SuperITMan commented:
About the warning in `stark-core`, it's a bug due to ng-packagr. There is a way to fix it but I'm don't think it's nice...

The issue is due to file `is-isin.validator.fn.ts` which contains to `for` loops with `let i = 0` inside:

```typescript
export function starkIsISIN(isin: string): boolean {
  // ...
  if (typeof isin === "string" && isinPattern.test(isin)) {
	let digits = "";

	for (let i = 0; i < lengthWithoutCheckDigit; i++) {
		digits += parseInt(isin[i], base).toString();
	}
	
	// ...
	
	for (let i = 0; i < digits.length; i++) {
		let digit: number = parseInt(digits[i], base);

		if (i % 2 === 0) {
			digit *= 2;
		}

		sum += floor(digit / modulo);
		sum += digit % modulo;
	}
	
	// ...
}

// ...
```

Which can be solved by changing the variable defined in the second `for` loop:

```typescript
export function starkIsISIN(isin: string): boolean {
  // ...
  if (typeof isin === "string" && isinPattern.test(isin)) {
	let digits = "";

	for (let i = 0; i < lengthWithoutCheckDigit; i++) {
		digits += parseInt(isin[i], base).toString();
	}
	
	// ...
	
	for (let j = 0; j < digits.length; j++) {
		let digit: number = parseInt(digits[j], base);

		if (j % 2 === 0) {
			digit *= 2;
		}

		sum += floor(digit / modulo);
		sum += digit % modulo;
	}
	
	// ...
}

// ...
```

What do you think about this @christophercr ?

The warnings in `stark-ui` are for the same reasons...

SuperITMan commented on CONTRIBUTING.md:
```diff
@@ -222,7 +222,7 @@ TODO add project structure details
 If you want to build Stark from source, you need to...
 
--   install NodeJS (6.0.0+) and npm (5.3.0+)
+-   install NodeJS (8.0.0+) and npm (5.6.0+)
```
You're right, it should be part of another PR

christophercr commented:
Ok, I see...  weird :s well then yes, I agree we should change the name of the variables to avoid duplication... it is easier to solve than I thought 😄

SuperITMan commented:
@christophercr I reverted changes in `CONTRIBUTING.md` and in karma config of showcase.
I added `ng-package.json` in testing subpackages.
I also fixed some warnings in stark-ui and stark-core but I decided to keep 3 of them because I think we'll loose in code clarity if we fix them...
See the "bad" code (we should change the `const column`: 
```typescript
if (globalFilter !== "%empty%") {
	// initially we take all the filter criteria as "unmet" criteria
	let unmetFilterCriteria: RegExp[] = this.getNormalizedFilterCriteria(globalFilter);

	for (const column of this.columns) {
		const displayedValue: string = column.getDisplayedValue(rowData).toString();
		// recalculate the "unmet" criteria again based on the remaining "unmet" criteria from the previous iteration
		unmetFilterCriteria = this.getUnmetFilterCriteria(displayedValue, unmetFilterCriteria);
	}

	matchFilter.push(unmetFilterCriteria.length === 0);
}

for (const column of this.columns) {
	if (column.filterValue) {
		const displayedValue: string = column.getDisplayedValue(rowData).toString();
		const filterCriteria: RegExp[] = this.getNormalizedFilterCriteria(column.filterValue);

		matchFilter.push(this.getUnmetFilterCriteria(displayedValue, filterCriteria).length === 0);
	}
}
```

Could you please have a look ? 😊

carlo-nomes commented:
@SuperITMan I think you can also close #1129 with this PR. 😄
